File: desktop/linux/src/native.cpp
// A tiny platform shim: cxx-qt-lib 0.10 does not wrap QFontDatabase or QStyleHints, so the
// couple of calls Crumb needs live here behind plain C++ functions.

#include "native.h"

#include <QtCore/QStringList>
#include <QtCore/QUrl>
#include <QtGui/QClipboard>
#include <QtGui/QDesktopServices>
#include <QtGui/QFontDatabase>
#include <QtGui/QGuiApplication>
#include <QtGui/QStyleHints>

int crumbAddApplicationFont(const QString &path) {
    return QFontDatabase::addApplicationFont(path);
}

QString crumbApplicationFontFamily(int id) {
    const QStringList families = QFontDatabase::applicationFontFamilies(id);
    return families.isEmpty() ? QString() : families.first();
}

bool crumbPrefersDark() {
#if QT_VERSION >= QT_VERSION_CHECK(6, 5, 0)
    // Unknown (no desktop hint) prefers dark, matching Crumb's default palette.
    return QGuiApplication::styleHints()->colorScheme() != Qt::ColorScheme::Light;
#else
    return true;
#endif
}

void crumbSetClipboardText(const QString &text) {
    QGuiApplication::clipboard()->setText(text);
}

bool crumbOpenUrl(const QString &url) {
    return QDesktopServices::openUrl(QUrl(url));
}

File: desktop/linux/src/native.rs
//! Rust bindings to `src/native.cpp` (fonts and colour scheme).

#[cxx_qt::bridge]
pub mod qobject {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qstring.h");
        type QString = cxx_qt_lib::QString;

        include!("native.h");

        #[rust_name = "add_application_font"]
        fn crumbAddApplicationFont(path: &QString) -> i32;

        #[rust_name = "application_font_family"]
        fn crumbApplicationFontFamily(id: i32) -> QString;

        #[rust_name = "prefers_dark"]
        fn crumbPrefersDark() -> bool;

        #[rust_name = "set_clipboard_text"]
        fn crumbSetClipboardText(text: &QString);

        #[rust_name = "open_url"]
        fn crumbOpenUrl(url: &QString) -> bool;
    }
}

pub use qobject::{
    add_application_font, application_font_family, open_url, prefers_dark, set_clipboard_text,
};
